docs/adr/adr_003_bundling_collection_data.md

```diff
@@ -0,0 +1,209 @@
+
+More alternatively, we could use event-driven architecture. The API would submit a request for the bundle to be built to a queue. A serverless function would process the request and return the response. In order to prevent multiple to build the same bundle multiple times, parallelism would have to be restricted, or we’d need to use a queue that supports partitioning based on collection name (or similar).
+- Good, because it all happens in one endpoint and easy to reason about
+- Good, because uploading/removing attachments or approving changes would not take longer
+- Bad, because a new specific container that has to be monitored, troubleshot, and deployed
+- Bad, because we have to introduce some complexity in order to prevent building several bundles in parallel
+
+### Option 5 - Build bundles on schedule
+A scheduled job iterates through the collections with the `attachment.bundle == true` flag and operates with the following logics:
+- If there is a bundle for the current collection timestamp, nothing to do.
+- If there isn’t, pull all attachments from storage, build a bundle with all current attachments, store it at `{folder}/attachments.zip` in cloud storage
+- (repeat for each bucket, including preview buckets)
+
+And creates a records bundle with all collections data (~6MB) for each bucket (`{folder}/changesets.zip`).
+
+*Note: the job that is currently in charge of publishing the highest timestamp to the Push notification server, could first execute this attachment bundling job, in order to make sure bundles will be available when clients will pull changesets.*
+
+This job would require GCloud credentials (Service Accounts credentials JSON file) and unlike `kinto-attachment` would be vendor-specific (ie. GCP Cloud storage using the `google-cloud-storage` Python library).
+
+We would have to duplicate some of the `kinto-attachment` configuration values (eg. `kinto.attachment.gcloud.bucket_name` and `kinto.attachment.folder`) in order to store the bundles along attachments (ie. in the right folder `{bucket id}/{collections id}`), so that the final bundle URL remains straightforward for clients (eg. `capabilities.attachments.base_url` + `bucket_id/collection_id /bundle.zip`)
+
+The clients would pull the records bundle on first sync, and the attachments bundle URL when `attachments.cacheAll()` is called. 
+- Good, because ultra simple
+- Good, because low cost
+- Good, because monitored as other remote settings cronjobs
+- Good, because it works easily for the preview bucket
+- Not so good, because we would introduce some coupling with kinto-attachment config (unless we mount the same RS server .ini config in the container and read conf from it) 
+- Not so good, because if we don’t couple it with the push timestamp cronjob, there could be a window of time where the bundle for a particular timestamp is not yet available
```

Comment:
We call this option "trivial" to implement, but these concerns don't exactly seem trivial.

Reply:
> because we would introduce some coupling with kinto-attachment config

I think this is mainly us pulling the same environment values, which is hopefully not too big of an issue.

> if we don’t couple it with the push timestamp cronjob, there could be a window of time where the bundle for a particular timestamp is not yet available

This is the normal CDN expiration concern, right? Maybe I'm minimizing this concern.  I think the negative scenario here is a client pulls yesterday's bundle and then has to pull some deltas for a few minutes. But if they had pulled the bundle just before publishing it's the same scenario. That doesn't seem too bad to me.

Reply:
> I think this is mainly us pulling the same environment values, which is hopefully not too big of an issue.

Yes, there is a little bit of duplication, but this remains trivial.

> if we don’t couple it with the push timestamp cronjob, there could be a window of time where the bundle for a particular timestamp is not yet available


Either we run the cronjob jobs together: ` python build_bundles.py && python sync_megaphone.py` (trivial)

Orwe accept that the few clients that will start a new profile right before the bundles are built, will pull an outdated bundle, and will have a few collections to synchronize after the extraction. (minor consequences)

We could add some details here if necessary